kotlinx-coroutines-core/jvm/test/FieldWalker.kt

```diff
@@ -7,109 +7,146 @@ package kotlinx.coroutines
-        val result = newSetFromMap<Any>(IdentityHashMap())
-        result.add(root)
+    private fun walkRefs(root: Any?): Map<Any, Ref> {
+        val visited = IdentityHashMap<Any, Ref>()
+        if (root == null) return visited
+        visited.put(root, Ref.RootRef)
         val stack = ArrayDeque<Any>()
         stack.addLast(root)
         while (stack.isNotEmpty()) {
             val element = stack.removeLast()
-            val type = element.javaClass
-            type.visit(element, result, stack)
+            try {
+                visit(element, visited, stack)
+            } catch (e: Exception) {
+                error("Failed to visit element ${showPath(element, visited)}: $e")
+            }
         }
-        return result
+        return visited
     }
 
-    private fun Class<*>.visit(
-        element: Any,
-        result: MutableSet<Any>,
-        stack: ArrayDeque<Any>
-    ) {
-        val fields = fields()
-        fields.forEach {
-            it.isAccessible = true
-            val value = it.get(element) ?: return@forEach
-            if (result.add(value)) {
-                stack.addLast(value)
+    private fun showPath(element: Any, visited: Map<Any, Ref>): String {
+        val path = ArrayList<String>()
+        var cur = element
+        while (true) {
+            val ref = visited[cur]!!
+            if (ref is Ref.RootRef) break
+            when (ref) {
+                is Ref.FieldRef -> {
+                    cur = ref.parent
+                    path += ".${ref.name}"
+                }
+                is Ref.ArrayRef -> {
+                    cur = ref.parent
+                    path += "[${ref.index}]"
+                }
             }
         }
+        path.reverse()
+        return path.joinToString("")
+    }
 
-        if (isArray && !componentType.isPrimitive) {
-            val array = element as Array<Any?>
-            array.filterNotNull().forEach {
-                if (result.add(it)) {
-                    stack.addLast(it)
+    private fun visit(element: Any, visited: IdentityHashMap<Any, Ref>, stack: ArrayDeque<Any>) {
+        val type = element.javaClass
+        when {
+            // Special code for arrays
+            type.isArray && !type.componentType.isPrimitive -> {
+                @Suppress("UNCHECKED_CAST")
+                val array = element as Array<Any?>
+                array.forEachIndexed { index, value ->
+                    push(value, visited, stack) { Ref.ArrayRef(element, index) }
+                }
+            }
+            // Special code for platform types that cannot be reflectively accessed on modern JDKs
+            type.name.startsWith("java.") && element is Collection<*> -> {
+                element.forEachIndexed { index, value ->
+                    push(value, visited, stack) { Ref.ArrayRef(element, index) }
+                }
+            }
+            type.name.startsWith("java.") && element is Map<*, *> -> {
```

Comment:
and everything else from the `jdk.base` we are going to fix/skip/workaround when we have a problem with it, right?

Reply:
Yes. That's the plan. It now fails with a detailed message when it encounters some JDK class we have not covered yet (it will exception when trying to make its fields accessible), so we'll be adding support here as needed.